### PR TITLE
[Ascend] qwen optimization

### PR DESCRIPTION
**Pull request by Liwansi**
## Motivation
related to #10337 

## Modifications
-bugfix:

1.memory bugfix(w8a8_int8.py): in previous code, both layer.w13_weight and layer.w2_weight occupied double memory. now we solve it.
2.Cache Management Operation(CMO) bugfix(common.py):in some circumstances(BS=1,2), deadlock situations may occur due to issues with stream sync and waiting. now we solve it.
3.eplb bugfix:The eplb index operator is also introduced when the map path is not set.

-optimization:

1.using triton high-performance fused OP named split_qkv_rmsnorm_rope, which involve modifications to both rotary_embedding.py and qwen3_moe.py.
2.using triton high-performance OP named l1_norm, which involve modifications to moe/topk.py
3.support PA into NPUGraph after some relevant software packages have been released, which involve modifications to npu_graph_runner.py and ascend_backend.py. see [Ascend]optimize Qwen3 on Ascend #10574
4.add moe-a2a-backend named ascend_fuseep, only support decode when pd disaggregation, integration of dispatch、gmm1、swiglu、gmm2、combine OP. Performance improved by 10% on Qwen3 235B.

## Accuracy Tests
- Qwen-32B
image

- Qwen-235B
image

## Benchmarking and Profiling

- Qwen-235B
> because some bugs before, we only have now performance after we changed.
image

> performance of fuseep:
fuseep
deepep

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).

### CODE REVIEW

File: python/sglang/srt/layers/moe/topk.py [hunk 1]
```diff
@@ -849,18 +850,26 @@ def select_experts(
...
-            gating_output=router_logits,
-            topk=top_k,
-            renormalize=renormalize,
-            correction_bias=correction_bias,
-        )
+        # assert (
+        #     num_token_non_padded is None
+        # ), "num_token_non_padded is not yet supported in fused_topk_native"
+        # assert expert_location_dispatch_info is None
+        # assert not apply_routed_scaling_factor_on_output, "Not implemented"
```
[the remark this answers is not in the document]
Comment (ping1jing2) [reply python/sglang/srt/layers/moe/topk.py]: remove these useless code

File: python/sglang/srt/utils/common.py [hunk 2]
```diff
@@ -605,44 +605,44 @@ def make_layers_non_pp(
...
     return cmo_stream
 
 
-def prepare_weight_cache(handle, cache):
+def set_cmo_stream(stream):
+    global cmo_stream
+    cmo_stream = stream
+
+
+def prepare_weight_cache(handle, cache, PREFETCH_MAX_SIZE = 1000000000):
```
Comment (ping1jing2) [inline python/sglang/srt/utils/common.py]: please add docstring to describe how to set `PREFETCH_MAX_SIZE`

File: python/sglang/srt/mem_cache/memory_pool.py [hunk 3]
```diff
@@ -1759,17 +1759,20 @@ def set_kv_buffer(
...
-            self.k_buffer[layer_id - self.start_layer].view(-1, 1, self.kv_lora_rank),
-            loc.view(-1, 1),
-            cache_k.view(-1, 1, self.kv_lora_rank),
-        )
-        torch_npu.npu_scatter_nd_update_(
-            self.v_buffer[layer_id - self.start_layer].view(
-                -1, 1, self.qk_rope_head_dim
+        import torch_npu
+        if loc.dtype != torch.int32:
+            loc = loc.to(torch.int32)
```
Comment (ping1jing2) [inline python/sglang/srt/mem_cache/memory_pool.py]: please remove these code and refer #11510

File: python/sglang/srt/layers/moe/topk.py [hunk 4]
```diff
@@ -68,6 +68,7 @@
 _is_cpu_amx_available = cpu_has_amx_support()
 _is_npu = is_npu()
 _use_aiter = get_bool_env_var("SGLANG_USE_AITER") and _is_hip
+_use_gating_topk_fused = get_bool_env_var("SGLANG_USE_GATING_TOPK_FUSED")
```
Comment (ping1jing2) [inline python/sglang/srt/layers/moe/topk.py]: please change it.
`_use_gating_topk_fused = get_bool_env_var("SGLANG_USE_GATING_TOPK_FUSED") and _is_npu`

File: python/sglang/srt/layers/rotary_embedding.py [hunk 5]
```diff
@@ -166,13 +169,27 @@ def _compute_cos_sin_cache(self) -> torch.Tensor:
         cache = torch.cat((cos, sin), dim=-1)
         return cache
 
+    def get_cos_sin_with_position(self, positions, layer_id):
```
Comment (ping1jing2) [inline python/sglang/srt/layers/rotary_embedding.py]: if this function is just for NPU, i think it would be better to name it like this:
`def _get_cos_sin_with_position(self, positions, layer_id):`

File: python/sglang/srt/mem_cache/memory_pool.py [hunk 6]
```diff
@@ -1759,17 +1759,18 @@ def set_kv_buffer(
...
 
-        torch_npu.npu_scatter_nd_update_(
-            self.k_buffer[layer_id - self.start_layer].view(-1, 1, self.kv_lora_rank),
-            loc.view(-1, 1),
-            cache_k.view(-1, 1, self.kv_lora_rank),
-        )
-        torch_npu.npu_scatter_nd_update_(
-            self.v_buffer[layer_id - self.start_layer].view(
-                -1, 1, self.qk_rope_head_dim
+        import torch_npu
```
Comment (ping1jing2) [inline python/sglang/srt/mem_cache/memory_pool.py]: please delete it because we have already done it at the beginning of this file.

File: python/sglang/srt/layers/moe/topk.py [hunk 7]
```diff
@@ -874,18 +875,20 @@ def select_experts(
...
+        if not _use_gating_topk_fused:
+            topk_weights, topk_ids = fused_topk_native(
+                hidden_states=hidden_states,
+                gating_output=router_logits,
+                topk=top_k,
+                renormalize=renormalize,
+                correction_bias=correction_bias,
+            )
+        else:
+            topk_weights, topk_ids, _ = torch_npu.npu_moe_gating_top_k_softmax(
```
Comment (iforgetmyname) [inline python/sglang/srt/layers/moe/topk.py]: this kernel has been introduced, pls pull the latest code

File: python/sglang/srt/layers/moe/topk.py [hunk 8]
```diff
@@ -68,6 +68,7 @@
 _is_cpu_amx_available = cpu_has_amx_support()
 _is_npu = is_npu()
 _use_aiter = get_bool_env_var("SGLANG_USE_AITER") and _is_hip
+_use_gating_topk_fused = get_bool_env_var("SGLANG_USE_GATING_TOPK_FUSED") and _is_npu
```
Comment (iforgetmyname) [inline python/sglang/srt/layers/moe/topk.py]: we don't need this env var, use this kernel by default if it's robust enough

File: python/sglang/srt/layers/rotary_embedding.py [hunk 9]
```diff
@@ -166,13 +169,27 @@ def _compute_cos_sin_cache(self) -> torch.Tensor:
         cache = torch.cat((cos, sin), dim=-1)
         return cache
 
+    def _get_cos_sin_with_position(self, positions, layer_id):
```
Comment (iforgetmyname) [inline python/sglang/srt/layers/rotary_embedding.py]: give this a more general impl that benefits gpu as well

File: python/sglang/srt/mem_cache/memory_pool.py
Comment (iforgetmyname) [inline python/sglang/srt/mem_cache/memory_pool.py:1755]: revert this change

File: python/sglang/srt/layers/moe/topk.py [hunk 10]
```diff
@@ -874,18 +875,20 @@ def select_experts(
...
-        assert expert_location_dispatch_info is None
         assert not apply_routed_scaling_factor_on_output, "Not implemented"
-        topk_weights, topk_ids = fused_topk_native(
-            hidden_states=hidden_states,
-            gating_output=router_logits,
-            topk=top_k,
-            renormalize=renormalize,
-            correction_bias=correction_bias,
-        )
+        if not _use_gating_topk_fused:
```
Comment (ping1jing2) [inline python/sglang/srt/layers/moe/topk.py]: same implementation:https://github.com/sgl-project/sglang/blob/main/python/sglang/srt/layers/moe/topk.py#L328

File: python/sglang/srt/models/qwen3_moe.py [hunk 11]
```diff
@@ -415,6 +418,7 @@ def forward_prepare(
                 and self.compatible_with_fused_kv_buffer
                 else None
             ),
+            layer_id=self.layer_id if _is_npu else None,
```
Comment (iforgetmyname) [inline python/sglang/srt/models/qwen3_moe.py]: revert this file

File: python/sglang/srt/utils/__init__.py [hunk 12]
```diff
@@ -1,2 +1,3 @@
 # Temporarily do this to avoid changing all imports in the repo
 from sglang.srt.utils.common import *
+from sglang.srt.utils.npu_common import *
```
Comment (iforgetmyname) [inline python/sglang/srt/utils/__init__.py]: no need to change in this pr

File: python/sglang/srt/layers/rotary_embedding.py [hunk 13]
```diff
@@ -220,6 +238,14 @@ def forward_npu(
             return self.forward_native(
                 positions, query, key, offsets, fused_set_kv_buffer_arg
             )
+        elif self.head_size == 128 and layer_id is not None:
```
Comment (ping1jing2) [inline python/sglang/srt/layers/rotary_embedding.py]: apply it both gpu and npu

File: python/sglang/srt/utils/common.py [hunk 14]
```diff
@@ -92,6 +91,7 @@
 
 from sglang.srt.environ import envs
 from sglang.srt.metrics.func_timer import enable_func_timer
+from sglang.srt.utils.npu_common import get_npu_compiler_config, get_npu_memory_capacity
```
Comment (iforgetmyname) [inline python/sglang/srt/utils/common.py]: revert this file, no need to change in this pr

File: python/sglang/srt/mem_cache/memory_pool_ascend.py [hunk 15]
```diff
@@ -0,0 +1,205 @@
...
+
+import torch
+import torch_npu
+
+from sglang.srt.constants import GPU_MEMORY_TYPE_KV_CACHE
+from sglang.srt.layers.radix_attention import RadixAttention
+from sglang.srt.mem_cache.memory_pool import MLATokenToKVPool, get_tensor_size_bytes
+
+
+class AscendMLAPagedTokenToKVPool(MLATokenToKVPool):
```
Comment (ping1jing2) [inline python/sglang/srt/mem_cache/memory_pool_ascend.py]: revert

File: python/sglang/srt/model_executor/model_runner.py [hunk 16]
```diff
@@ -108,6 +99,7 @@
     ReqToTokenPool,
     SWAKVPool,
 )
+from sglang.srt.mem_cache.memory_pool_ascend import AscendMLAPagedTokenToKVPool
```
Comment (ping1jing2) [inline python/sglang/srt/model_executor/model_runner.py]: revert

File: python/sglang/srt/layers/rotary_embedding.py (hunk 13 above)
Comment (ping1jing2) [reply python/sglang/srt/layers/rotary_embedding.py]: revert

File: python/sglang/srt/models/qwen3_moe.py [hunk 17]
```diff
@@ -398,6 +427,12 @@ def forward_prepare(
     ):
         if hidden_states.shape[0] == 0:
             return hidden_states, forward_batch, None
+        if _is_npu:
```
Comment (chenxu140) [inline python/sglang/srt/models/qwen3_moe.py]: create forward_prepare_native func

File: python/sglang/srt/layers/moe/ep_moe/layer.py [hunk 18]
```diff
@@ -155,6 +155,16 @@ def forward(
                 hidden_states,
                 topk_output,
             )
+        if self.deepep_mode.is_low_latency() and get_bool_env_var("ENABLE_ASCEND_FUSED_DEEPEP"):
+            return self.forward_npu_fused_deepep(
```
Comment (iforgetmyname) [inline python/sglang/srt/layers/moe/ep_moe/layer.py]: implement `fused_deepep_moe` as a standalone `moe_a2a_backend` instead

Take these into consideration:
- add `DispatchOutputChecker` type
- how to escape `run_moe_core` and `dispatch`

File: python/sglang/srt/models/qwen3_moe.py [hunk 19]
```diff
@@ -390,14 +398,37 @@ def op_core(self, state):
             state.pop("attn_intermediate_state")
         )
 
-    def forward_prepare(
+    def forward_prepare_npu(
```
Comment (ssshinigami) [inline python/sglang/srt/models/qwen3_moe.py:403]: Please try to avoid this model change and create fusion pass for this kernel
We have supported fusion pass manager in this PR
https://github.com/sgl-project/sglang/pull/11104

**Review by ssshinigami — CHANGES_REQUESTED** [review]

File: python/sglang/srt/layers/moe/ep_moe/layer.py [hunk 20]
```diff
@@ -492,9 +496,142 @@ def _forward_ll(dispatch_output: DeepEPLLDispatchOutput):
...
+            gmm1_permuted_weight_scale=self.w13_weight_scale,
+            gmm2_weight=self.w2_weight,
+            gmm2_weight_scale=self.w2_weight_scale,
+        ).hidden_state
+
+    def release_weight_cache(self, weight: torch.Tensor):
+        # .contiguous() introduces additional memory overhead and needs to be released using resize_(0)
+        origin_weight = weight.data.transpose(1, 2)
+        new_weight = origin_weight.contiguous()
+        origin_weight.untyped_storage().resize_(0)
```
Comment (ssshinigami) [inline python/sglang/srt/layers/moe/ep_moe/layer.py:472]: not needed this workaround, we have fix in this PR https://github.com/sgl-project/sglang/pull/11984
commit https://github.com/sgl-project/sglang/pull/11984/commits/ff1a3bb1e5514dcb8284bbe8b978b368665ff421

File: python/sglang/srt/layers/quantization/w8a8_int8.py [hunk 21]
```diff
@@ -1001,15 +1007,23 @@ def create_weights(
         layer.register_parameter("w2_weight_offset", w2_weight_offset)
         set_weight_attrs(w2_weight_offset, extra_weight_attrs)
 
+    def release_weight_cache(self, weight: torch.Tensor):
+        # .contiguous() introduces additional memory overhead and needs to be released using resize_(0)
+        origin_weight = weight.data.transpose(1, 2)
+        new_weight = origin_weight.contiguous()
+        origin_weight.untyped_storage().resize_(0)
```
Comment (ssshinigami) [inline python/sglang/srt/layers/quantization/w8a8_int8.py:1085]: not needed this workaround, we have fix in this PR https://github.com/sgl-project/sglang/pull/11984
commit https://github.com/sgl-project/sglang/pull/11984/commits/ff1a3bb1e5514dcb8284bbe8b978b368665ff421

**Review by ssshinigami — CHANGES_REQUESTED** [review]

File: python/sglang/srt/layers/moe/ep_moe/layer.py (hunk 20 above)
Comment (iforgetmyname) [reply python/sglang/srt/layers/moe/ep_moe/layer.py:472]: i can't really appreciate anything that does fix this issue there

using `flatten` is just another workaround from my view

**Review by iforgetmyname — APPROVED** [review]

File: python/sglang/srt/layers/moe/ep_moe/layer.py [hunk 22]
```diff
@@ -492,9 +496,142 @@ def _forward_ll(dispatch_output: DeepEPLLDispatchOutput):
             raise ValueError(f"Not Supported DeepEP format {dispatch_output.format}")
 
 
+class NpuFuseEPMoE(DeepEPMoE):
```
[the remark this answers is not in the document]
Comment (iforgetmyname) [reply python/sglang/srt/layers/moe/ep_moe/layer.py:418]: will be solved in #13359
